### DISCUSSION TITLE
How to write an ipv6 udp echo client?

### DISCUSSION THREAD

**Discussion by yinkaisheng** [agronholm/anyio#793] [Q&A]
I use Windows 10, python 3.12.6 and anyio 4.6.
I wrote the code:
```python
import sys
import socket
import anyio

async def server():
    print('server starts')
    async with await anyio.create_udp_socket(family=socket.AF_INET6, local_host='::', local_port=25000) as udp:
        async for packet, (host, port) in udp:
            print(f'recv from {host} {port}: {packet}')
            print(f'send to {host} {port}: {packet}')
            await udp.sendto(packet, host, port)

async def client():
    host, port = '::1', 25000
    async with await anyio.create_udp_socket(family=socket.AF_INET6) as udp:
        laddr = udp.extra(anyio.abc.SocketAttribute.local_address)
        print(f'client {laddr}')
        for i in range(4):
            data = f'hello {i}'.encode()
            print(f'send to {host} {port}: {data}')
            await udp.sendto(data, host, port)
            resp, raddr = await udp.receive()
            print(f'recv from {raddr}: {resp}')

if __name__ == '__main__':
    func = server if sys.argv[1] == 'server' else client
    anyio.run(func)
```
When a client runs, the server doesn't receive the data.
![image](https://github.com/user-attachments/assets/94a4a596-8f98-4021-80e1-f726616f9a40)

**Comment by smurfix** [comment]
Meh. Works on Linux, unsurprisingly.

What happens if you use IPv4?

**Reply by yinkaisheng** [reply]
It doesn't work either when I use IPv4. Works in WSL.
If I use stdlib socket to send data, both ipv4 and ipv6 servers can receive data. It may be the client's issue.

![image](https://github.com/user-attachments/assets/7266ceae-365d-421b-8ea3-d0218cea9e8e)

**Reply by agronholm** [reply]
I also couldn't get it working on Windows. But AnyIO has tests to make sure that UDP works, and these tests are passing on Windows. I'm stumped.

**Reply by agronholm** [reply]
It works if I use the Ethernet ip address instead of ::1.

**Reply by agronholm** [reply]
It also works fine on the `trio` backend, even against localhost.

**Reply by agronholm** [reply]
It's worth noting that we're skipping asyncio ipv6 UDP tests "due to a known bug" (https://github.com/python/cpython/issues/83329). At least there used to be a problem with IOCP support around this, but I don't know if this is still the case. At least the upstream issue has been closed.

**Reply by agronholm** [reply]
Relevant: https://github.com/agronholm/anyio/blob/master/tests/test_sockets.py#L115

**Reply by agronholm** [reply]
If I comment out the use of that fixture, the IPv6 UDP tests hang.

**Reply by yinkaisheng** [reply]
OK.
